### ISSUE TITLE
CryptoHelper::arrayCompare - leaks info about arrays

### ISSUE THREAD

**Issue by meg23** [ESAPI/esapi-java-legacy#246]
_From [noloa...@gmail.com](https://code.google.com/u/114558122492435650190/) on August 06, 2011 04:53:07_

Oops... three cases where the comment does not hold. At minimum, it might be appropriate to spin based on case 3's min/avg/max length.

if ( b1 == b2 ) {
  return true;
}
if ( b1 == null || b2 == null ) {
  return (b1 == b2);
}
if ( b1.length != b2.length ) {
  return false;
}

int result = 0;
// Make sure to go through ALL the bytes. We use the fact that if
// you XOR any bit stream with itself the result will be all 0 bits,
// which in turn yields 0 for the result.
for(int i = 0; i < b1.length; i++) {
  // XOR the 2 current bytes and then OR with the outstanding result.
  result |= (b1[i] ^ b2[i]);
}
return (result == 0) ? true : false;

_Original issue: http://code.google.com/p/owasp-esapi-java/issues/detail?id=239_

**Comment by kwwall** [comment]
Corrected title because name of method was arrayCompare(), not compareArray().

This method was only present because at the time, we supported JDK 1.6, which meant there were versions of that JDK where MessageDigest.isEqual() had timing side-channel attacks. This was supposedly fixed in some version of JDK 6 and since the new ESAPI release only supports JDK 7 and later, this method has been deprecated and changed to simply be implemented as this:

```
	public static boolean arrayCompare(byte[] b1, byte[] b2) {
            return java.security.MessageDigest.isEqual(b1, b2);
	}
```

**Comment by kwwall** [comment]
Closed via PR #464
